### PR TITLE
Add x509 renew

### PR DESCRIPTION
**Pull request by thomasmmitchell**
`x509 renew`. Added.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -2297,6 +2303,119 @@ The following options are recognized:
...
+			rand.Reader, cert.Certificate, ca.Certificate,
+			cert.PrivateKey.Public(), ca.PrivateKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		newCert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			panic("Could not parse certificate we just made. We so STUPID!!!")
```
Comment (jhunt) [inline main.go]: Let's tone down the panic here...

File: main.go [hunk 2]
```diff
@@ -2297,6 +2303,119 @@ The following options are recognized:
...
+  -i, --signed-by   Path in the Vault where the CA certificate
+                    (and signing key) can be found.  If this is not
+                    provided, a sibling secret named 'ca' will used
+                    if it exists.
+
+  -t, --ttl         How long the new certificate will be valid
+                    for.  Specified in units h (hours), m (months)
+                    d (days) or y (years).  1m = 30d and 1y = 365d
+                    Defaults to same duration used by the certificate
+                    starting at the time of the renew.
```
Comment (jhunt) [inline main.go]: "starting at the time of the renew" is confusing here.  I think you can rephrase as "Defaults to the original TTL used to issue the certificate"

File: main.go (hunk 1 above)
Comment (dennisjbell) [reply main.go]: what, don't like the UHF reference??!!

File: main.go (hunk 2 above)
Comment (dennisjbell) [reply main.go]: k

Comment (dennisjbell) [reply main.go]: used `Defaults to the last TTL used to issue or renew the certificate` as that's more accurate.

**Comment by dennisjbell** [comment]
reissue added, review changes performed

File: main.go [hunk 3]
```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
...
+		Summary: "Reissue X.509 Certificates and Certificate Authorities",
+		Usage:   "safe x509 reissue [OPTIONS] path/to/certificate",
+		Description: `
+Reissues an X.509 Certificate with a new key.
+
+The following options are recognized:
+
+  -b, --bits  N      RSA key strength, in bits.  The only valid
+                     arguments are 1024 (highly discouraged),
+                     2048 and 4096.  Defaults to 4096
```
Comment (jhunt) [inline main.go]: This defaults to the last value used to (re)issue the certificate.

File: main.go [hunk 4]
```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
...
+
+  -i, --signed-by   Path in the Vault where the CA certificate
+                    (and signing key) can be found.  If this is not
+                    provided, a sibling secret named 'ca' will used
+                    if it exists.
```
Comment (jhunt) [inline main.go]: Might want to note that this _should_ be the same as the original signing CA, but need not be.

File: main.go [hunk 5]
```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
...
+		// Get signing key bit length
+		var bits int
+		if opt.X509.Reissue.Bits != 0 {
+			bits = opt.X509.Reissue.Bits
+			if bits != 1024 && bits != 2048 && bits != 4096 {
+				return fmt.Errorf("Bits must be one of 1024, 2048 or 4096")
+			}
+		} else {
+			bits = cert.PrivateKey.N.BitLen()
+		}
```
Comment (jhunt) [inline main.go]: minor nit; you don't actually need the `bits` var here.  You can just update opt.X509.Reissue.Bits

File: main.go [hunk 6]
```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
...
+		if caPath != args[0] {
+			err = ca.SaveTo(v, caPath, false)
 			if err != nil {
 				return err
 			}
 		}
 
-		s, err := cert.Secret(opt.SkipIfExists)
+		err = cert.SaveTo(v, args[0], false)
```
Comment (jhunt) [inline main.go]: This brings up an interesting point.

Should we complain if someone tries to `safe x509 reissue --skip-if-exists`?

File: main.go [hunk 7]
```diff
@@ -2275,25 +2288,192 @@ The following options are recognized:
...
+                    provided, a sibling secret named 'ca' will used
+                    if it exists.
+
+  -t, --ttl         How long the new certificate will be valid
+                    for.  Specified in units h (hours), m (months)
+                    d (days) or y (years).  1m = 30d and 1y = 365d
+                    Defaults to the last TTL used to issue or renew
+                    the certificate.
+
+	`,
```
Comment (jhunt) [inline main.go]: Is this a trailing `\t`?

File: main.go [hunk 8]
```diff
@@ -2625,3 +2805,45 @@ func recursively(cmd string, args ...string) bool {
 	y = strings.TrimSpace(y)
 	return y == "y" || y == "yes"
 }
+
+func findSigningCa(v *vault.Vault, cert *vault.X509, certPath string, signPath string) (*vault.X509, string, error) {
```
Comment (jhunt) [inline main.go]: Should this move into the `vault` module proper?

i.e.

```
vault.FindSigningCA(cert, path, ca)
```

File: vault/x509.go [hunk 9]
```diff
@@ -458,6 +458,20 @@ func (x X509) Secret(skipIfExists bool) (*Secret, error) {
...
+	if err != nil {
+		return err
+	}
+
+	err = v.Write(path, s)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (jhunt) [inline vault/x509.go]: You can just return v.Write()

**Pull request merged**